### DISCUSSION TITLE
docker/系统升级导致数据丢失解决方法

### DISCUSSION THREAD

**Discussion by LetMeDecay** [MoeNetwork/Tieba-Cloud-Sign#266] [交流]
这两天ubuntu20.04提示可以升级，我就dist-upgrade一下，然后容器就寄了😅启动之后就提示重装，研究了下怎么解决
先进入compose配置文件夹

### 数据未被删除
```
docker exec -it sign-db-1 /bin/bash
mariadb -uroot -pjanejane123456
show databases;
use tiebacloud;
select * from tc_tieba limit 10;
```
只要这一步能查到用户，创建install.lock文件之后就能读取之前的数据了
先把compose文件里面这两行注释去掉
```
    volumes:                                        # 如果重新创建或更新镜像希望保留之前的所有信息，请解除此两行注释
      - ./install.lock:/var/www/setup/install.lock  # 在当前目录下自行新建文件install.lock，内容为：1
```
然后创建一个lock文件
```
touch install.lock
```
最后重建容器就行了
```
docker compose up -d
```

### 数据已删除(未测试，不确定是否可用)
先备份mysql文件夹，防止折腾失败
```
cp -r mysql bak
```
然后根据重置数据库的时间找到具体的binlog文件
```
cd mysql;
ll;
```
通常来说日期靠后并且比较大的那个文件就是
然后把下面的时间调整到重置之前，binlog名字改成具体的文件
```
docker exec -it sign-db-1 /bin/bash
cd /var/lib/mysql;
mysqlbinlog --no-defaults --start-datetime='2024:07:19 08:00:00' mysql-bin.000004 | mysql -uroot -pjanejane123456
```
重置之后再走未删除的方式就行了

**Comment by n0099** [comment]
> 然后把下面的时间调整到重置之前，binlog名字改成具体的文件
> [...]
> mysqlbinlog --no-defaults --start-datetime='2024:07:19 08:00:00' mysql-bin.000004 | mysql -uroot -pjanejane123456

这本质上是利用binlog固有特征实现的`point in time恢复` https://dev.mysql.com/doc/refman/8.4/en/point-in-time-recovery-binlog.html 其要求先有完整备份
但实际上可以直接修改binlog中每条目中的几个位把`DELETE`改成`INSERT`来实现另一个意义上的`point in time回滚`（rollback不完全是recovery，但最终效果可能一样） https://github.com/lefred/MyUndelete

**Reply by LetMeDecay** [reply]
正常来说不手贱重新初始化，按第一个流程就能解决，也就不需要binlog这种方法了
更复杂的方法完全没必要，有那折腾时间还不如重建站点，再进行一次绑定

**Reply by n0099** [reply]
> 还不如重建站点，再进行一次绑定

您让 https://n0099.net/tc 的几百个僵尸用户在`430作战计划`期间四叶服被`UCSB加州大学圣芭芭拉分校ctf中级高手`苏琦神攻克后拖裤泄露所有注册用户绑定的BDUSS导致我当时不得不退出登录这些bduss后去重新绑定吧

**Reply by LetMeDecay** [reply]
😅抽象
